kv/dist_sender.go

```diff
@@ -491,154 +606,63 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
 // This may temporarily adjust the request headers, so the client.Call
 // must not be used concurrently until Send has returned.
 func (ds *DistSender) Send(_ context.Context, call client.Call) {
+	args := call.Args
+	finalReply := call.Reply
+	endKey := args.Header().EndKey
 
-	// TODO: Refactor this method into more manageable pieces.
 	// Verify permissions.
 	if err := ds.verifyPermissions(call.Args); err != nil {
 		call.Reply.Header().SetGoError(err)
 		return
 	}
 
-	// Retry logic for lookup of range by key and RPCs to range replicas.
-	retryOpts := ds.rpcRetryOptions
-	retryOpts.Tag = "routing " + call.Method().String() + " rpc"
-
-	var descNext *proto.RangeDescriptor
-	args := call.Args
-	reply := call.Reply
-	endKey := args.Header().EndKey
-
 	// In the event that timestamp isn't set and read consistency isn't
 	// required, set the timestamp using the local clock.
 	if args.Header().ReadConsistency == proto.INCONSISTENT && args.Header().Timestamp.Equal(proto.ZeroTimestamp) {
+		// Make sure that after the call, args hasn't changed.
+		defer func() {
+			args.Header().Timestamp = proto.ZeroTimestamp
+		}()
 		args.Header().Timestamp = ds.clock.Now()
 	}
 
-	// If this is an InternalPushTxn, set ignoreIntents option as
-	// necessary. This prevents a potential infinite loop; see the
-	// comments in proto.InternalRangeLookupRequest.
-	options := lookupOptions{}
-	if pushArgs, ok := args.(*proto.InternalPushTxnRequest); ok {
-		options.ignoreIntents = pushArgs.RangeLookup
+	// If this is a bounded request, we will change its bound as we receive
```

Comment:
The bound-changing may happen repeatedly. I tried, but it looked more awkward than this.